### PR TITLE
Return mode even if single value (#15714)

### LINKED ISSUE #15714: Mode should be defined if the Series consist of only one object
#### Code Sample

```python
>>> import pandas.core.algorithms as algos
>>> algos.mode([5])
Series([], dtype: int64) 
```
#### Problem description

Currently the `mode` of a series is not defined if the series consist of only one object. This limits the use of the mode as an aggregation method.

Compare this with the behaviour of other libraries, e.g. `scipy`:

```python
>>> from scipy import stats
>>> scipy.stats.mode([5])
(array([ 5.]), array([ 1.]))
```

#### Expected Output

```python
0    5
dtype: int64
``` 
#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 2.7.10.final.0
python-bits: 64
OS: Darwin
OS-release: 14.5.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.19.2
nose: 1.3.7
pip: 9.0.1
setuptools: 34.3.1
Cython: None
numpy: 1.8.0rc1
scipy: 0.13.0b1
statsmodels: None
xarray: None
IPython: 5.1.0
sphinx: 1.4.8
patsy: None
dateutil: 2.6.0
pytz: 2016.10
blosc: None
bottleneck: None
tables: None
numexpr: None
matplotlib: 1.3.1
openpyxl: 2.3.0-b1
xlrd: 1.0.0
xlwt: None
xlsxwriter: 0.9.3
lxml: 3.7.3
bs4: None
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: 1.1.6
pymysql: None
psycopg2: 2.6.2 (dt dec pq3 ext lo64)
jinja2: 2.9.5
boto: 2.42.0
pandas_datareader: None
</details>

### PR DESCRIPTION
**Pull request by buyology**
- [X] Please accept this PR which closes #15714 
- [X] Tests added — `test_mode_single`
- [X] Passes ``git diff upstream/master | flake8 --diff`` (however: it complains about some non-changed rows in hashtable_func_helper.pxi.in)

### CODE REVIEW

**Comment by jreback** [comment]
@buyology looks like a couple of failing tests. I think we tests this condition of a single object for multiple types, so need to change it in each.

**Comment by buyology** [comment]
Tests updated

File: pandas/core/frame.py [hunk 1]
```diff
@@ -5168,9 +5168,8 @@ def _get_agg_axis(self, axis_num):
 
     def mode(self, axis=0, numeric_only=False):
         """
-        Gets the mode(s) of each element along the axis selected. Empty if
```
Comment (jreback) [inline pandas/core/frame.py]: in this (or followup), we should make this a shared doc-string :>

File: pandas/tests/frame/test_analytics.py [hunk 2]
```diff
@@ -837,20 +837,9 @@ def test_mode(self):
...
-        # and also when not empty
-        df.loc[1, "A"] = 0
-        df.loc[4, "B"] = df.loc[3, "B"]
-        df.loc[5, "C"] = 'e'
-        exp = pd.DataFrame({"A": pd.Series([0], dtype=df["A"].dtype),
-                            "B": pd.Series([df.loc[3, "B"]],
-                                           dtype=df["B"].dtype),
-                            "C": pd.Series(['e'], dtype=df["C"].dtype)})
-
+        exp = pd.DataFrame({"A": pd.Series(np.arange(6, dtype='int64'), dtype=df["A"].dtype),
```
Comment (jreback) [inline pandas/tests/frame/test_analytics.py]: you have flake issues

``git diff master | flake8 --diff``

see [here](http://pandas-docs.github.io/pandas-docs-travis/contributing.html#python-pep8)

File: pandas/tests/test_algos.py [hunk 3]
```diff
@@ -1256,12 +1256,30 @@ def test_no_mode(self):
...
+            tm.assert_series_equal(algos.mode(s), exp)
+
+        exp = Series([1], dtype=np.int)
         tm.assert_series_equal(algos.mode([1]), exp)
 
-        exp = Series([], dtype=np.object)
+        exp = Series(['a', 'b', 'c'], dtype=np.object)
         tm.assert_series_equal(algos.mode(['a', 'b', 'c']), exp)
 
+    def test_mode_single(self):
```
Comment (jreback) [inline pandas/tests/test_algos.py]: this looks duplicated

File: pandas/tests/test_algos.py [hunk 4]
```diff
@@ -1256,12 +1256,30 @@ def test_no_mode(self):
         exp = Series([], dtype=np.float64)
         tm.assert_series_equal(algos.mode([]), exp)
 
-        exp = Series([], dtype=np.int)
+    def test_mode_single(self):
+        exp_single = [1]
```
Comment (jreback) [inline pandas/tests/test_algos.py]: add the issue number as a comment

**Review by jreback — APPROVED** [review]
just some some minor comments. ping when green.

File: pandas/core/frame.py (hunk 1 above)
Comment (buyology) [inline pandas/core/frame.py]: @jreback — not sure how to do this, I reckon this is to reference the same-ish doc string in all relevant places?

File: pandas/tests/frame/test_analytics.py (hunk 2 above)
Comment (buyology) [inline pandas/tests/frame/test_analytics.py]: fixed all but the ones in the `hashtable_func_helper.pxi.in` — for some reason it's complaining about lines that I haven't even touched / don't show up in the diffs here on GH

File: pandas/core/frame.py (hunk 1 above)
Comment (jreback) [inline pandas/core/frame.py]: yes you woould move ``def mode`` to ``generic.py`` (leaving a stub here and in series for an ``@Appender`` doc-string (look at virtually any other doc-string, e.g. ``fillna``)

can do this as a follow up though (or here, up 2 u)

File: pandas/tests/test_algos.py [hunk 5]
```diff
@@ -1256,10 +1256,20 @@ def test_no_mode(self):
...
+    # GH 15714
+    def test_mode_single(self):
+        exp_single = [1]
+        data_single = [1]
+
+        for dt in np.typecodes['AllInteger'] + np.typecodes['Float']:
+            s = Series(data_single, dtype=dt)
+            exp = Series(exp_single, dtype=dt)
+            tm.assert_series_equal(algos.mode(s), exp)
+
```
Comment (jreback) [inline pandas/tests/test_algos.py]: can you add the result for a double-but-same value

e.g.

``data = [1, 1]`` (exp should be the same)

**Comment by jreback** [comment]
can you add a whatsnew note (Bug Fix section). small comment, otherwise lgtm. ping on green.

**Comment by jreback** [comment]
can you update

**Comment by jreback** [comment]
thanks @buyology 

nice PR!

keep em coming!

**Pull request closed without merging**